### PR TITLE
gzip integration tests

### PR DESCRIPTION
**Pull request by gargnitingoogle**
### Description
This adds read-flow integration tests for the gzip read-compressed changes added in previous PR https://github.com/GoogleCloudPlatform/gcsfuse/pull/1255 .

Tests added (2*5 = 10 tests)
* Two types of operations each
  * List, Stat and full-file read
* On five types of objects each
  * Object with text content, uploaded with gzip compression with content-encoding: gzip, and cache-control: no-transform
  * Object with text content, uploaded with gzip compression with content-encoding: gzip, and cache-control: ''
  * Object with gzip content, uploaded with content-encoding: ''
  * Object with gzip content, uploaded with gzip compression (i.e. doubly compressed) with content-encoding: gzip, and cache-control: no-transform
  * Object with gzip content, uploaded with gzip compression (i.e. doubly compressed) with content-encoding: gzip, and cache-control: ''

This change is dependent on https://github.com/GoogleCloudPlatform/gcsfuse/pull/1255 .

Write-flow tests are added in https://github.com/GoogleCloudPlatform/gcsfuse/pull/1262. 

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - Ran all the tests for testbucket and mountedDirectory flags.(status:SUCCESS)
2. Unit tests - NA
3. Integration tests - Run all the tests for testbucket and mountedDirectory flags.(status:SUCCESS)

### CODE REVIEW

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [inline tools/integration_tests/gzip/gzip_test.go:43]: On enabling `--implicit-dirs` I get the following error log:

```
...
2023/08/03 04:04:47 /tmp/gcsfuse_readwrite_test_971925334/bin/gcsfuse --sequential-read-size-mb=1 --implicit-dirs --debug_gcs --debug_fs --debug_fuse --log-file=/tmp/gcsfuse_readwrite_test_971925334/gcsfuse.log --log-format=text gargnitin-gzip-test-bucket /tmp/gcsfuse_readwrite_test_971925334/mnt
2023/08/03 04:04:47 mountGcsfuse: cannot mount gcsfuse: exit status 1
```

On checking `/tmp/gcsfuse_readwrite_test_971925334/gcsfuse.log`, there is nothing in the log:
```
gargnitin@gargnitin:~/work/cloud/storage/client/gcsfuse/src/gcsfuse$ ls -la /tmp/gcsfuse_readwrite_test_971925334/gcsfuse.log
-rw-r--r-- 1 gargnitin primarygroup 278 Aug  3 04:04 /tmp/gcsfuse_readwrite_test_971925334/gcsfuse.log
gargnitin@gargnitin:~/work/cloud/storage/client/gcsfuse/src/gcsfuse$ cat /tmp/gcsfuse_readwrite_test_971925334/gcsfuse.log
/tmp/gcsfuse_readwrite_test_971925334/bin/gcsfuse --sequential-read-size-mb=1 --implicit-dirs --debug_gcs --debug_fs --debug_fuse --log-file=/tmp/gcsfuse_readwrite_test_971925334/gcsfuse.log --log-format=text gargnitin-gzip-test-bucket /tmp/gcsfuse_readwrite_test_971925334/mnt
```

Comment copied from https://github.com/GoogleCloudPlatform/gcsfuse/pull/1255#discussion_r1282198052 .

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [inline tools/integration_tests/gzip/gzip_test.go:139]: `--enable-storage-client-library=false` fails with the same error as in  https://github.com/GoogleCloudPlatform/gcsfuse/pull/1256/commits/3ac6df6dc005e94f3a2c34964d7d76747bb60f03#r1282722070

**Comment by gargnitingoogle** [comment]
Resolve this comment only after https://github.com/GoogleCloudPlatform/gcsfuse/pull/1255/files is merged.

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/gzip_test.go]: This is now fixed.

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/gzip_test.go]: This is now fixed.

**Review by Tulsishah — COMMENTED** [review]
Also please check once if in sync with other tests it is working properly with both the flags.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:89]: File handle is already open during CreateTemp, so it doesn't make sense to open it again. All the other precautions are already taken here.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:134]: Done.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:70]: Done.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:64]: Done.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:48]: Done.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:107]: DOne.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:114]: Done

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:191]: Done.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:156]: It will not panic. It'll always return error and the stdout buffer and let the caller deal with it.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:172]: > Let's try to use gcloud from now

Done.

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/gzip_test.go:150]: @Tulsishah  This leaves behind the question of where to put the setup/destroy calls in that case .

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/read_gzip_test.go:85]: The reason I don't want to use that function is that it opens the file, reads a chunk and closes it. If I use that function here, it'll force me to open and close the file n times adding unnecessary runtime overhead to this test.

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/gzip_test.go:150]: Discussed this offline with @Tulsishah . Will do this.

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/gzip_test.go:153]: Discussed this offline with @Tulsishah . This is no needed.

File: tools/integration_tests/gzip/gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/gzip_test.go:154]: Yeah. Will add as discussed.

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (Tulsishah) [reply tools/integration_tests/gzip/helpers/helpers.go]: Also Can't we execute script and return stderr in [this](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/tools/integration_tests/util/setup/setup.go#L50) function itself. It is easy to understand and also not much code is require?

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (Tulsishah) [inline tools/integration_tests/gzip/helpers/helpers.go:119]: Let's keep all the functions which are common like executeToolCommandf, executeGcloudCommandf in  util/setup and GetGcsObjectSize,... in util/operations

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (Tulsishah) [inline tools/integration_tests/gzip/read_gzip_test.go:37]: use StatFile function which you have created in operations.

File: tools/integration_tests/run_tests_mounted_directory.sh
Comment (Tulsishah) [inline tools/integration_tests/run_tests_mounted_directory.sh:282]: Left one line at the end.

File: tools/integration_tests/util/operations/file_operations.go
Comment (Tulsishah) [inline tools/integration_tests/util/operations/file_operations.go:301]: Can we use filePath1 and filePath2 name?

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (Tulsishah) [inline tools/integration_tests/gzip/read_gzip_test.go:101]: Please write comment before each test. like [this](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/tools/integration_tests/operations/copy_dir_test.go#L126)

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (Tulsishah) [inline tools/integration_tests/gzip/helpers/helpers.go:122]: nit: keep one line between variable initialisation and function calling.

var stdout bytes.Buffer
var stderr bytes.Buffer

err := cmd.Run()
	if err != nil {
		return stdout.Bytes(), fmt.Errorf("failed command '%s': %v, %s", cmdArgs, err, stderr.String())
	}

**Review by Tulsishah — COMMENTED** [review]
Please add testing details, what you have done in the PR description. 
e.g.
```
1. Manual - Run all the tests for testbucket and mountedDirectory flags.(status:SUCCESS)
2. Unit tests - NA
3. Integration tests - Run all the tests for testbucket and mountedDirectory flags.(status:SUCCESS)
```

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/read_gzip_test.go:101]: The name of the test function is self-explanatory. The referenced comment lists out files, directories copied by the test. There is no such list for the gzip tests.

File: tools/integration_tests/run_tests_mounted_directory.sh
Comment (gargnitingoogle) [reply tools/integration_tests/run_tests_mounted_directory.sh:282]: I don't get this comment. There is no line at the end.

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/read_gzip_test.go:101]: I've added explainatory comments on the generic*Test functions to explain what they do.

File: tools/integration_tests/run_tests_mounted_directory.sh
Comment (Tulsishah) [reply tools/integration_tests/run_tests_mounted_directory.sh:278]: I mean cursor should be in new line in the file.

**Review by Tulsishah — COMMENTED** [review]
Please wait for @sethiay approval also before merging the PR.

File: tools/integration_tests/run_tests_mounted_directory.sh
Comment (gargnitingoogle) [reply tools/integration_tests/run_tests_mounted_directory.sh]: Done.

**Comment by gargnitingoogle** [comment]
@sethiay please review this PR.

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (sethiay) [reply tools/integration_tests/gzip/read_gzip_test.go:94]: Given n is small here, I would also prefer to use that function.

Comment (gargnitingoogle) [reply tools/integration_tests/gzip/read_gzip_test.go]: Done.

File: tools/integration_tests/util/operations/file_operations.go
Comment (sethiay) [inline tools/integration_tests/util/operations/file_operations.go:303]: I believe this function is not used in the current tests ? If yes, it might be good to add it in a separate PR ?

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:43]: No, that's allowed and supported.

File: tools/integration_tests/util/operations/file_operations.go
Comment (gargnitingoogle) [reply tools/integration_tests/util/operations/file_operations.go:303]: My intention was to use it in the full-file read tests to diff the content in mounted directory with the content in the gcs bucket. I'll add those calls.

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/read_gzip_test.go:125]: I purposefully kept those two as separate tests because of
1. For granularity of testing, and
2. The size test checks the output of stat call or `os.Stat` as opposed to the read size which tests the NewReader call.

There is of course redundancy between the two, so I'll merge the two tests into one.

File: tools/integration_tests/util/operations/file_operations.go
Comment (gargnitingoogle) [reply tools/integration_tests/util/operations/file_operations.go:303]: Done. Being used in the gzip tests now.

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/read_gzip_test.go]: Done, renamed.

File: tools/integration_tests/gzip/read_gzip_test.go
Comment (gargnitingoogle) [reply tools/integration_tests/gzip/read_gzip_test.go]: Merged into one.

**Comment by gargnitingoogle** [comment]
@Tulsishah @sethiay please approve if no more comments.

**Review by Tulsishah — APPROVED** [review]

File: tools/integration_tests/gzip/helpers/helpers.go
Comment (sethiay) [inline tools/integration_tests/gzip/helpers/helpers.go:125]: is there any reason why we are not using gcloud/gsutil here to download gzip file?

Comment (gargnitingoogle) [reply tools/integration_tests/gzip/helpers/helpers.go:125]: Yes, that was my first instinct as well, but gcloud/gsutil don't support reading back a gzip object as compressed. I tried with all arguments that I knew of, but none works, and they always return the decompressed content on `gsutil cp` or `gcloud storage cp`. So, I had to fallback to go-storage-library, same as I have done in the actual implementation.

Comment (sethiay) [reply tools/integration_tests/gzip/helpers/helpers.go:125]: Got it, I think it's worth mentioning that as comment.

**Review by sethiay — APPROVED** [review]

**Pull request merged**
